### PR TITLE
Strengthten upgrade process

### PR DESCRIPTION
**Pull request by Hlavtox**
| Questions         | Answers
| ------------------| -------------------------------------------------------
| Description?      | This should reduce the amount of potential failure points for upgrade. It creates tables only if they don't exist. It adds columns to tables only if they don't exist. If they do, they are updated. Common case when people backport some feature from newer versions, or upgrade failed mid-point somehow.
| Type?             | bug fix
| BC breaks?        | no
| Deprecations?     | no
| Fixed ticket?     | 
| Sponsor company   | TRENDO s.r.o.
| How to test?      | Tests green.

### What I changed
- New script add_column that adds a column if it doesn't exist OR modifies it, if it already does.
- I made sure that all `ALTER TABLE ADD` are using this script.
- I checked that all `CREATE TABLE` are conditioned with `IF EXISTS`.
- I checked that all `DROP TABLE` are conditioned with `IF EXISTS`.
- I checked that all `DROP COLUMN` are using our script `drop_column_if_exists`.

### What it improves
- When the upgrade accidentaly runs twice, @kpodemski somehow reproduced it.
- When people backport some features from newer versions and already have some extra column. Or when some module added it.
- (Maybe) When an upgrade failed and people just restored the files.

Ping @Quetzacoalt91

### CODE REVIEW

**Review by matks — COMMENTED** [review]
Looks good and a safe way to do it 👍 

Wdyt @M0rgan01 @Quetzacoalt91 @ga-devfront ?

**Comment by Hlavtox** [comment]
Thanks. :-))) I extended the PR description.

**Comment by M0rgan01** [comment]
@Hlavtox > When the upgrade accidentaly runs twice, @kpodemski somehow reproduced it.

How could this happen ?

**Comment by Hlavtox** [comment]
@M0rgan01 No idea, ask @kpodemski, but he won't know, just happened 🤣

File: upgrade/sql/8.0.0.sql [hunk 1]
```diff
@@ -203,14 +203,14 @@ ALTER TABLE `PREFIX_webservice_permission` MODIFY COLUMN `method` ENUM(
...
 
-ALTER TABLE `PREFIX_product` ADD `unit_price` decimal(20, 6) NOT NULL DEFAULT '0.000000' AFTER `unity`;
-ALTER TABLE `PREFIX_product_shop` ADD `unit_price` decimal(20, 6) NOT NULL DEFAULT '0.000000' AFTER `unity`;
+/* PHP:add_column('product', 'unit_price', "decimal(20, 6) NOT NULL DEFAULT '0.000000' AFTER `unity`"); */;
+/* PHP:add_column('product_shop', 'unit_price', "decimal(20, 6) NOT NULL DEFAULT '0.000000' AFTER `unity`"); */;
 
 UPDATE `PREFIX_product` SET `unit_price` = IF (`unit_price_ratio` != 0, `price` / `unit_price_ratio`, 0);
 UPDATE `PREFIX_product_shop` SET `unit_price` = IF (`unit_price_ratio` != 0, `price` / `unit_price_ratio`, 0);
 
+/* PHP:add_column('feature_flag', 'stability', "VARCHAR(64) DEFAULT 'beta' NOT NULL"); */;
```
Comment (M0rgan01) [inline upgrade/sql/8.0.0.sql:212]: This doesn't seem to work, from CI :
a536f11bad6852d3615699fa5dd7eccdfbabc4f2

Comment (M0rgan01) [reply upgrade/sql/8.0.0.sql:212]: I test locally to see the state of the DB

Comment (M0rgan01) [reply upgrade/sql/8.0.0.sql:212]: I tried without the double quotes, and it works... the handling of calls in CoreUpgrader::runPhpQuery seems to be a problem? In any case, collums are not created

Comment (Hlavtox) [reply upgrade/sql/8.0.0.sql]: @M0rgan01 Hi, could you please check it and see why the stability query doesn't work? :( I used escaped quotes everywhere, but it did not fix it.

Comment (M0rgan01) [reply upgrade/sql/8.0.0.sql]: I'm trying to find a solution

Comment (M0rgan01) [reply upgrade/sql/8.0.0.sql]: @Hlavtox Good news, https://github.com/PrestaShop/autoupgrade/pull/874 solves our problem

**Comment by M0rgan01** [comment]
waiting for https://github.com/PrestaShop/autoupgrade/pull/874 merge

**Comment by Hlavtox** [comment]
@M0rgan01 @Quetzacoalt91 Rebased, let's see what tests say :-)

**Review by M0rgan01 — APPROVED** [review]

**Comment by Hlavtox** [comment]
Yaaaaay 🥳

**Review by ga-devfront — APPROVED** [review]

**Review by AureRita — APPROVED** [review]
Hi @Hlavtox 

Thank you for your PR, I tested it and it seems to works as you can see : 

[recording(294).webm](https://github.com/user-attachments/assets/f40a46e7-c4b3-4e55-8032-d34aa13c0e9d)

Tested with UI from : 
1.7.7.8 to 8.1.7
1.7.7.8 to 1.7.8.11 
8.0.5 to 8.1.7 
8.1.7 to 9.0.0
8.0.5 to 9.0.0

Tested with SeamlessUpgradeToolbox : 
1.7.0.1 to 1.7.2.0 
1.7.0.1 to 1.7.3.0 
1.7.5.1 to 1.7.6.0
1.7.5.1 to 1.7.6.6
1.7.5.1 to 1.7.7.0 
1.7.5.1 to 1.7.8.0 
1.7.5.1 to 8.0.0
1.7.5.1 to 8.1.0 
8.0.4 to 9.0.0 

With this test I see this part that disapear with the upgrade : 

image

That why i tested to put a note with an order. I think it disapear because we don't put often a note on order at the installation. That's why I tested it on the video 

Because the PR seems to works as expected, It's QA :heavy_check_mark: 

Thank you

**Comment by rdy4ever** [comment]
Late to the party, but thank you for this. I have a bad habit of backporting security patches and new features instead of performing a full upgrade if possible (due to how many times a full upgrade messed up something on my side) - so this PR is a godsend. Sometimes back-porting makes it easier to test, debug and later deploy to prod each new patch while still having a sane, readable git diff as opposed to a full upgrade that makes it hard to identify exactly what went wrong. So again, thank you @Hlavtox 🙏

**Comment by Hlavtox** [comment]
@rdy4ever Ha, no worries, you are welcome @rdy4ever. :-) My reasoning is exactly same as yours, I backport all of the fixes required, so I can avoid doing upgrades if not needed - they are a lot of work, mostly on prep. This makes it smoooooth sail.

**Pull request merged**